text/3389-manifest-lint.md

```diff
@@ -0,0 +1,269 @@
+`[lints.<target>]` or `[target.<target>.lints]` but
+- There isn't a defined use case for this yet besides having support for `cfg(feature = "clippy")` or
+  which does not seem high enough priority to design
+  around.
+- `[lints.<target>]` runs into ambiguity issues around what is a `<target>`
+  entry vs a `<lint>` entry in the `[lints]` table.
+- We have not yet defined semantics for sharing something like this across a
+  workspace
+
+Instead of using workspace inheritance for `[lint]`, we could make it
+workspace-level configuration, like `[patch]` which is automatically applied to
+all workspace members.  However, `[patch]` and friends are because they affect
+the resolver / `Cargo.toml` and so they can only operate at the workspace
+level.  `[lints]` is more like `[dependencies]` in being something that applies
+at the package level but we want shared across workspaces.
+
+Instead of traditional workspace inheritance where there is a single value to
+inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
+which defines presets and the user could do `lints.<preset> = true`.  The user
+could then name them as they wish to avoid collision with rustc lints.
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Rust
+- [cargo cranky](https://github.com/ericseppanen/cargo-cranky)
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
```

Comment:
Would it work for cargo to just hard code what lint tool goes to what CLI? 

The problem is then dealing with the potential future state of third-party lint tools.  If those were opted into via cargo instead of a source-level attribute, we might be able to get the information we need to direct the lints to the appropriate CLI (or ignore as needed)

Reply:
Couldn't cargo always pass all lints defined in the `[lint]` section to all tools? If a tool is not used and therefore not registered with `rustc`, `rustc` will just ignore that lint and that it was passed to it. Or am I misunderstanding the problem? :thinking:

Reply:
I've not touched that side of things so I am unsure if that would be a problem or not.

ehuss also brought up the possibility of using this to minimize which lints are included in fingerprinting for reducing recompiles for lint changes (e.g. adding clippy lints doesn't cause `cargo build` to rebuild).  Its mentioned in the RFC just after this paragraph.